### ISSUE TITLE
[Redis] Prioritized tasks are always enqueued as highest priority

### ISSUE THREAD

**Issue by flbraun** [celery/celery#8967]
# Checklist

- [x] I have verified that the issue exists against the `main` branch of Celery.
- [x] This has already been asked to the [discussions forum](https://github.com/celery/celery/discussions) first.
- [x] I have read the relevant section in the
  [contribution guide](https://docs.celeryq.dev/en/main/contributing.html#other-bugs)
  on reporting bugs.
- [x] I have checked the [issues list](https://github.com/celery/celery/issues?q=is%3Aissue+label%3A%22Issue+Type%3A+Bug+Report%22+-label%3A%22Category%3A+Documentation%22)
  for similar or identical bug reports.
- [x] I have checked the [pull requests list](https://github.com/celery/celery/pulls?q=is%3Apr+label%3A%22PR+Type%3A+Bugfix%22+-label%3A%22Category%3A+Documentation%22)
  for existing proposed fixes.
- [x] I have checked the [commit log](https://github.com/celery/celery/commits/main)
  to find out if the bug was already fixed in the main branch.
- [x] I have included all related issues and possible duplicate issues
  in this issue (If there are none, check this box anyway).

## Mandatory Debugging Information

- [x] I have included the output of ``celery -A proj report`` in the issue.
    (if you are not able to do this, then at least specify the Celery
     version affected).
- [x] I have verified that the issue exists against the `main` branch of Celery.
- [x] I have included the contents of ``pip freeze`` in the issue.
- [x] I have included all the versions of all the external dependencies required
  to reproduce this bug.

## Optional Debugging Information

- [ ] I have tried reproducing the issue on more than one Python version
  and/or implementation.
- [ ] I have tried reproducing the issue on more than one message broker and/or
  result backend.
- [ ] I have tried reproducing the issue on more than one version of the message
  broker and/or result backend.
- [ ] I have tried reproducing the issue on more than one operating system.
- [ ] I have tried reproducing the issue on more than one workers pool.
- [ ] I have tried reproducing the issue with autoscaling, retries,
  ETA/Countdown & rate limits disabled.
- [ ] I have tried reproducing the issue after downgrading
  and/or upgrading Celery and its dependencies.

#### Related Issues

- None

#### Possible Duplicates

- None

## Environment & Settings

**Celery version**: 5.3.6

<details>
<summary><b><code>celery report</code> Output:</b></summary>
<p>

```
software -> celery:5.3.6 (emerald-rush) kombu:5.3.5 py:3.11.7
            billiard:4.2.0 redis:5.0.3
platform -> system:Linux arch:64bit
            kernel version:6.5.0-27-generic imp:CPython
loader   -> celery.loaders.app.AppLoader
settings -> transport:redis results:disabled

broker_url: 'redis://redis.local.tonie.cloud:6379/8'
deprecated_settings: None
imports: ['tasks']
broker_transport_options: 
    'priority_steps': [0, 25, 50, 75, 100],
    'queue_order_strategy': 'priority',
    'sep': ':'}
task_default_priority: 50

```

</p>
</details>

# Steps to Reproduce

## Required Dependencies

- **Minimal Python Version**: N/A or Unknown
- **Minimal Celery Version**: N/A or Unknown
- **Minimal Kombu Version**: N/A or Unknown
- **Minimal Broker Version**: N/A or Unknown
- **Minimal Result Backend Version**: N/A or Unknown
- **Minimal OS and/or Kernel Version**: N/A or Unknown
- **Minimal Broker Client Version**: N/A or Unknown
- **Minimal Result Backend Client Version**: N/A or Unknown

### Python Packages

<details>
<summary><b><code>pip freeze</code> Output:</b></summary>
<p>

```
```

</p>
</details>

### Other Dependencies

<details>
<p>
N/A
</p>
</details>

## Minimally Reproducible Test Case

<details>
<p>

```python
#
# app.py
#
import os

from celery import Celery


app = Celery('tasks', broker=f"redis://{os.environ['REDIS_HOST']}:{os.environ['REDIS_PORT']}/8")
app.conf.imports = ['tasks']
app.conf.broker_transport_options = {
    'priority_steps': [0, 25, 50, 75, 100],
    'sep': ':',
    'queue_order_strategy': 'priority',
}
app.conf.task_default_priority = 50

#
# tasks.py
#
import time

from prio import app


@app.task
def default():
    print('default')
    time.sleep(1)


@app.task(priority=75)
def high():
    print('high')
    time.sleep(1)


if __name__ == '__main__':
    for i in range(10):
        default.delay()
    high.delay()

```
Run `python tasks.py` first to enqueue some tasks, then run `celery -A prio:app worker --concurrency 1` to consume them.
</p>
</details>

# Expected Behavior

Tasks are enqueued in the Redis key matching the tasks' priority.

# Actual Behavior

When spawning a couple of tasks (see Minimally Reproducible Test Case) with a default priority of 50 and one task with a higher priority of 75, *all* tasks are enqueued into the highest priority key (in this case `celery`, which I assume is an alias for `celery:0`) instead of `celery:50` and `celery:75`, respectively.
You can see this behavior using `redis-cli MONITOR` (task body truncated):
```
1713192234.524321 [8 172.20.0.22:44290] "LPUSH" "celery" "{... \"priority\": 50, ...}"
1713192234.525032 [8 172.20.0.22:44290] "LPUSH" "celery" "{... \"priority\": 50, ...}"
1713192234.525566 [8 172.20.0.22:44290] "LPUSH" "celery" "{... \"priority\": 50, ...}"
1713192234.526095 [8 172.20.0.22:44290] "LPUSH" "celery" "{... \"priority\": 50, ...}"
1713192234.526651 [8 172.20.0.22:44290] "LPUSH" "celery" "{... \"priority\": 50, ...}"
1713192234.527272 [8 172.20.0.22:44290] "LPUSH" "celery" "{... \"priority\": 50, ...}"
1713192234.527900 [8 172.20.0.22:44290] "LPUSH" "celery" "{... \"priority\": 50, ...}"
1713192234.528452 [8 172.20.0.22:44290] "LPUSH" "celery" "{... \"priority\": 50, ...}"
1713192234.528958 [8 172.20.0.22:44290] "LPUSH" "celery" "{... \"priority\": 50, ...}"
1713192234.529502 [8 172.20.0.22:44290] "LPUSH" "celery" "{... \"priority\": 50, ...}"
1713192234.530302 [8 172.20.0.22:44290] "LPUSH" "celery" "{... \"priority\": 75, ...}"
```
The worker polls tasks using the correct Redis command:
```
1713192822.408074 [8 172.20.0.22:59782] "BRPOP" "celery" "celery:25" "celery:50" "celery:75" "celery:100" "1"
```
But since all tasks ended up in the same key, all queued tasks are effectively worked on in the order they were enqueued, *NOT* by their priority.

I also tested this using `'priority_steps': [1, 25, 50, 75, 100]` because I was wary of seeing the Redis key `celery` instead of the expected `celery:0`, but it yields the exact same results. All tasks are pushed into `celery:1`, just as they were pushed into `celery` before.

**Comment by zzJinux** [comment]
This bugged me as well until I figured out the cause. While searching for a solution, I just ran into the same dead end.

TL;DR: **Celery (more precisely, kombu) won't let you use priority values other than `[0,9]`.** It's an expected behavior (what?).

The priority value of a message is determined by: https://github.com/celery/kombu/blob/v5.4.2/kombu/transport/virtual/base.py#L847-L865

For your case, the priority value `50` is clamped to `9`.

By https://github.com/celery/kombu/blob/v5.4.2/kombu/transport/redis.py#L1015-L1017, `9` then changes to `0`.

And finally, since the `0` priority is treated specially, the plain queue name is chosen instead: https://github.com/celery/kombu/blob/v5.4.2/kombu/transport/redis.py#L1009-L1013

I've tested with your configuration and seen it in action to prove it.
